# List of tasks and ideas to be tested

- Try a fresh installation of phyloscannerR to see if it works:
devtools::install on the phyloscannerR directory

- Try to install R, phyloscannerR and all its dependencies, using conda

- Distribute the scripts and tool's scripts as intended by shiver & phyloscanner & HIV-PhyloTSI (now all scripts are merged within bin) -> checked: should not be done!

- Clarify if --x-iqtree is an option now

- Create custom HIV_DistanceNormalisationOverGenome.csv and compare it with an available file

- Compare IVA with SPADES

- Double check MAF associated sctipts

- Optimise HPC resources for all process if needed

- Clarify excluded windows: 9470,9720,9480,9730,9490,9740 from windows250_VR_norms_oneline.txt (Tanya)

- Clarify the option --read-names-2. It does not exists anymore! (Tanya)

- Clafiry the differences in testpatstats.csv: Prop.gp.1 - Prop.gp.215 (Tanya)

- Check on 4refs_HXB2_ABCD.fasta vs 2refs_HXB2_C.BW.fasta for --alignment-of-other-refs. Should I expect any dramatic difference?